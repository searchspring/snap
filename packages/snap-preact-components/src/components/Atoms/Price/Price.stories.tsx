import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY, Markdown } from '@storybook/blocks';

import { Price, PriceProps } from './Price';
import { componentArgs, highlightedCode } from '../../../utilities';
import Readme from '../Price/readme.md';

export default {
	title: 'Atoms/Price',
	component: Price,
	tags: ['autodocs'],
	parameters: {
		docs: {
			page: () => (
				<div>
					<Markdown
						options={{
							overrides: {
								code: highlightedCode,
							},
						}}
					>
						{Readme}
					</Markdown>
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	argTypes: {
		value: {
			description: 'Numeric value to be formatted',
			type: { required: true },
			table: {
				type: {
					summary: 'number',
				},
			},
			control: { type: 'number' },
		},
		symbol: {
			description: 'Currency symbol',
			defaultValue: '$',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: '$' },
			},
			control: { type: 'text' },
		},
		symbolAfter: {
			description: 'Place currency symbol after the value',
			defaultValue: false,
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: { type: 'boolean' },
		},
		decimalPlaces: {
			description: 'Number of decimal places',
			defaultValue: 2,
			table: {
				type: {
					summary: 'number',
				},
				defaultValue: { summary: 2 },
			},
			control: { type: 'number' },
		},
		padDecimalPlaces: {
			description: 'Pad decimal places with zeros',
			defaultValue: true,
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: true },
			},
			control: { type: 'boolean' },
		},
		thousandsSeparator: {
			description: 'Character used to separate thousands',
			defaultValue: ',',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: ',' },
			},
			control: { type: 'text' },
		},
		decimalSeparator: {
			description: 'Character used to separate decimal values',
			defaultValue: '.',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: '.' },
			},
			control: { type: 'text' },
		},
		lineThrough: {
			description: 'Add line through styling',
			defaultValue: false,
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: {
				type: 'boolean',
			},
		},
		raw: {
			description: 'Returns raw value without wrapping DOM node',
			defaultValue: false,
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: { type: 'boolean' },
		},
		...componentArgs,
	},
};

export const Default = (args: PriceProps) => <Price {...args} />;
Default.args = {
	value: 1099.99,
};

export const lineThrough = (args: PriceProps) => <Price {...args} />;
lineThrough.args = {
	value: 1199.99,
	lineThrough: true,
};

export const CustomCurrency = (args: PriceProps) => <Price {...args} />;
CustomCurrency.args = {
	value: 999.99,
	symbol: ' â‚¬',
	thousandsSeparator: '.',
	decimalSeparator: ',',
	symbolAfter: true,
};
