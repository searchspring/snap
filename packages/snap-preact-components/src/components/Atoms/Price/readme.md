## Overlay

Renders an span with a formatted number. Essentially a pricing shortcut for the FormattedNumber Component.

## Additional Info

Uses the Currency function from the Snap ToolBox (https://github.com/searchspring/snap-toolbox/blob/main/filters/src/currency.ts) 

## Usage

Default
```jsx
    <Price value={1099.99} />
```

Custom Pricing
```jsx
    <Price value={1099.99} symbol=' â‚¬' lineThrough={true} thousandsSeparator='.' decimalSeparator=',' symbolAfter={true}/>
```

