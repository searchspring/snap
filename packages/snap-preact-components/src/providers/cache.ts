export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
