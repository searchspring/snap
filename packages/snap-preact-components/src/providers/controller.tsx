import { jsx } from '@emotion/react';
import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractController } from '@searchspring/snap-controller';

const ControllerContext = createContext<null | AbstractController>(null);

export const ControllerProvider = ({ children, controller }: { children: ComponentChildren; controller: AbstractController }) => {
	return <ControllerContext.Provider value={controller}>{children}</ControllerContext.Provider>;
};

export const useController = () => useContext(ControllerContext);

export const withController = (Component: ComponentType) => (props: any) => {
	return <Component controller={useController()} {...props} />;
};
