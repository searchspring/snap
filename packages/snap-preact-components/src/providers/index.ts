export * from './store';
export * from './theme';
