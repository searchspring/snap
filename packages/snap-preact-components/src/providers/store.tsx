import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractStore } from '@searchspring/snap-store-mobx';

const StoreContext = createContext<null | AbstractStore>(null);

export const StoreProvider = ({ children, store }: { children: ComponentChildren; store: AbstractStore }) => {
	return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => useContext(StoreContext);

export function withStore<C extends ComponentType>(Component: C): C {
	return ((props: any) => <StoreContext.Consumer>{(store) => <Component {...props} store={store} />}</StoreContext.Consumer>) as C;
}
