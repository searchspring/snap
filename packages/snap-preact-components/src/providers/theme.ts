import { ThemeProvider as EmotionTheme } from '@emotion/react';
export { css, useTheme, withTheme } from '@emotion/react';

interface ThemeProviderProps {
	theme: Partial<Theme> | ((outerTheme: Theme) => Theme);
	children?: any;
}
interface ThemeProvider {
	(props: ThemeProviderProps): React.ReactElement;
}
export const ThemeProvider: ThemeProvider = EmotionTheme;

export const defaultTheme: Theme = {
	colors: {
		primary: '#3A23AD',
		secondary: '#00cee1',
		hover: '#f8f6fd',
		text: {
			secondary: '#ffffff',
		},
		message: {
			error: '#cc1212',
			warning: '#ecaa15',
			info: '#4c3ce2',
		},
	},
	components: {},
};
export interface Theme {
	variables?: {
		[key: string]: unknown;
	};
	colors?: {
		primary: string;
		secondary: string;
		hover?: string;
		text?: {
			secondary: string;
		};
		message?: {
			error: string;
			warning: string;
			info: string;
		};
	};
	components?: any;
	namedComponents?: any;
}
