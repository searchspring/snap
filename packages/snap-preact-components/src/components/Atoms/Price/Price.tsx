/** @jsx jsx */
import { h, Fragment } from 'preact';

import { filters } from '@searchspring/snap-toolbox';
import { jsx, css } from '@emotion/react';
import classnames from 'classnames';

import { Theme, useTheme, CacheProvider } from '../../../providers';
import { FormattedNumberProps } from '../FormattedNumber/FormattedNumber';
import { StylingCSS } from '../../../types';
import { mergeProps } from '../../../utilities';

const CSS = {
	price: ({ theme }: Partial<PriceProps>) =>
		css({
			color: theme?.colors?.primary,
			'&.ss__price--strike': {
				textDecoration: 'line-through',
				color: 'initial',
			},
		}),
};

export function Price(properties: PriceProps): JSX.Element {
	const globalTheme: Theme = useTheme();
	const defaultProps: Partial<PriceProps> = {
		symbol: '$',
		decimalPlaces: 2,
		padDecimalPlaces: true,
		thousandsSeparator: ',',
		decimalSeparator: '.',
		symbolAfter: false,
		lineThrough: false,
	};

	const props = mergeProps('price', globalTheme, defaultProps, properties);

	const {
		lineThrough,
		value,
		symbol,
		decimalPlaces,
		padDecimalPlaces,
		thousandsSeparator,
		decimalSeparator,
		symbolAfter,
		raw,
		disableStyles,
		className,
		style,
		styleScript,
	} = props;

	let formattedPrice: string | undefined;
	if (value) {
		formattedPrice = filters.currency(value, {
			symbol,
			decimalPlaces,
			padDecimalPlaces,
			thousandsSeparator,
			decimalSeparator,
			symbolAfter,
		});
	}

	const styling: { css?: StylingCSS } = {};
	const stylingProps = props;

	if (styleScript && !disableStyles) {
		styling.css = [styleScript(stylingProps), style];
	} else if (!disableStyles) {
		styling.css = [CSS.price(stylingProps), style];
	} else if (style) {
		styling.css = [style];
	}
	if (formattedPrice) {
		return raw ? (
			<Fragment>{formattedPrice}</Fragment>
		) : (
			<CacheProvider>
				<span {...styling} className={classnames('ss__price', { 'ss__price--strike': lineThrough }, className)}>
					{formattedPrice}
				</span>
			</CacheProvider>
		);
	} else {
		return <Fragment></Fragment>;
	}
}

export interface PriceProps extends Omit<FormattedNumberProps, 'value'> {
	value?: number;
	lineThrough?: boolean;
}
