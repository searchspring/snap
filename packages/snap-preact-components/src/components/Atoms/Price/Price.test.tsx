import { h } from 'preact';
import { render } from '@testing-library/preact';

import { Price } from './Price';
import { ThemeProvider } from '../../../providers';

describe('Price Component', () => {
	it('renders', () => {
		const rendered = render(<Price value={1099.99} />);
		const priceElement = rendered.container.querySelector('.ss__price');
		expect(priceElement).toBeInTheDocument();
	});

	it('has a line-through', () => {
		const args = {
			value: 1099.99,
			lineThrough: true,
		};
		const rendered = render(<Price {...args} />);
		const priceElement = rendered.container.querySelector('.ss__price')!;
		const styles = getComputedStyle(priceElement);
		expect(styles.textDecoration).toBe('line-through');
	});

	it('has default custom options', () => {
		const args = {
			value: 1099.99,
		};
		const rendered = render(<Price {...args} />);
		const priceElement = rendered.container.querySelector('.ss__price');
		const priceText = priceElement?.textContent;
		expect(priceText).toBe('$1,099.99');
	});

	it('has custom options', () => {
		const args = {
			value: 1099.99,
			symbol: ' £',
			decimalPlaces: 2,
			thousandsSeparator: '.',
			decimalSeparator: '.',
			symbolAfter: true,
		};
		const rendered = render(<Price {...args} />);
		const priceElement = rendered.container.querySelector('.ss__price');
		const symbolElement = rendered.container.querySelector('.ss__price__symbol');

		const priceText = priceElement?.textContent;
		expect(priceText).toBe('1.099.99 £');

		expect(symbolElement).toBeDefined();
		expect(symbolElement?.textContent).toBe(args.symbol);
	});

	it('renders with additional style using prop', () => {
		const style = {
			padding: '20px',
		};

		const rendered = render(<Price value={100} style={style} />);
		const priceElement = rendered.container.querySelector('.ss__price')!;
		const styles = getComputedStyle(priceElement);

		expect(styles.padding).toBe(style.padding);
	});

	it('can disable styling', () => {
		const args = {
			value: 1099.99,
			disableStyles: true,
		};
		const rendered = render(<Price {...args} />);
		const priceElement = rendered.container.querySelector('.ss__price');
		expect(priceElement?.className).not.toMatch(/formatted-/);
	});

	it('can custom className', () => {
		const args = {
			value: 1099.99,
			className: 'custom-class',
		};
		const rendered = render(<Price {...args} />);
		const priceElement = rendered.container.querySelector('.ss__price');
		expect(priceElement?.classList).toContain(args.className);
	});

	it('is themeable with ThemeProvider', () => {
		const globalTheme = {
			components: {
				price: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				<Price value={1099.99} />
			</ThemeProvider>
		);

		const priceElement = rendered.container.querySelector('.ss__price');
		expect(priceElement).toHaveClass(globalTheme.components.price.className);
	});

	it('is themeable with theme prop', () => {
		const propTheme = {
			components: {
				price: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(<Price value={1099.99} theme={propTheme} />);

		const priceElement = rendered.container.querySelector('.ss__price');
		expect(priceElement).toHaveClass(propTheme.components.price.className);
	});

	it('is themeable and theme prop overrides ThemeProvider', () => {
		const globalTheme = {
			components: {
				price: {
					className: 'global-class',
				},
			},
		};

		const propTheme = {
			components: {
				price: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				<Price value={1099.99} theme={propTheme} />
			</ThemeProvider>
		);

		const priceElement = rendered.container.querySelector('.ss__price');
		expect(priceElement).toHaveClass(propTheme.components.price.className);
		expect(priceElement).not.toHaveClass(globalTheme.components.price.className);
	});
});
