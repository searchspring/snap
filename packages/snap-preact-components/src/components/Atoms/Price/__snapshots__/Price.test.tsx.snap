// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Component can custom className 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price custom-class emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component can disable styling 1`] = `
<DocumentFragment>
  <span
    class="ss__price"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component has a line-through 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price ss__price--strike emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component has custom options 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price emotion-0"
  >
    1.099.99 Â£
  </span>
</DocumentFragment>
`;

exports[`Price Component has default custom options 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component is themeable and theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price theme-class emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price theme-class emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component is themeable with theme prop 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price theme-class emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component renders 1`] = `
<DocumentFragment>
  .emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price emotion-0"
  >
    $1,099.99
  </span>
</DocumentFragment>
`;

exports[`Price Component renders with additional style using prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 20px;
}

.emotion-0.ss__price--strike {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  color: initial;
}

<span
    class="ss__price emotion-0"
  >
    $100.00
  </span>
</DocumentFragment>
`;
